### ISSUE TITLE
Potential Performance Problem with Nancy.Routing.Route.Invoke

### ISSUE THREAD

**Issue by Luo-Liang** [NancyFx/Nancy#2623]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of Nancy
- [x] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [x] I have searched [open](https://github.com/NancyFx/Nancy/issues) and [closed](https://github.com/NancyFx/Nancy/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
Hi there! We may have found a potential problem with performance in Nancy.Routing.Route.Invoke. The use of `TaskCompletionSource` is really slowing down the performance of Nancy when tested with the sample `Nancy.Demo.Hosting.Aspnet` by as much as 10x. This happens both in Debug and Release mode.

### Steps to Reproduce
Run `wrk`, the workload generator with the following parameters in original code.
`-c 256 -t 1 - d 60 http://localhost:59017/user/liang` (256 connections, 1 thread and run for 60 seconds)
and
`-c 1 -t 1 -d 60 http://localhost:59017/user/liang` (1 connection, 1 thread and run for 60 seconds).
`wrk` reports `132` and `270` requests/sec.
The `http://localhost:59017/user/liang` is simply the site serviced by Nancy.
If we change the source code of `Route.cs` from

````
public override Task<object>  Invoke(DynamicDictionary parameters, CancellationToken cancellationToken)
{
    var task = this.Action.Invoke(parameters, cancellationToken);
    var tcs = new TaskCompletionSource<object
    task.ContinueWith(t => tcs.SetResult(t.Result), TaskContinuationOptions.OnlyOnRanToCompletion);
    task.ContinueWith(t => tcs.SetException(t.Exception.InnerExceptions), TaskContinuationOptions.OnlyOnFaulted);
    task.ContinueWith(t => tcs.SetCanceled(), TaskContinuationOptions.OnlyOnCanceled);
    return tcs.Task;
}
````
to simply
````
public override async Task<object> Invoke(DynamicDictionary parameters, CancellationToken cancellationToken)
{
    var task = this.Action.Invoke(parameters, cancellationToken);
    var configuredTask = task.ConfigureAwait(false);
    return await configuredTask;
 }
````
We get `864` and `1129` requests/second respectively with the same configuration on the same machine.

We understand there must be some engineering concern behind this, but the 10x slowdown may be a concern, at least for the common path.

### System Configuration
Intel XEON CPU E5-2673 with 7GB memory and Windows 10 anniversary update and .NET 4.5.

- Nancy version:  Checked out on 14/11/2016
- Nancy host
  - [x] ASP.NET
  - [ ] OWIN
  - [ ] Self-Hosted
  - [ ] Other:
- Other Nancy packages and versions: None
- Environment (Operating system, version and so on): Windows 10 Enterprise
- .NET Framework version: 4.5
- Additional information: Thank you for your patience :)

**Comment by Luo-Liang** [comment]
bump! Any ideas with this? Thanks!

**Comment by grumpydev** [comment]
Not sure why this was missed with the async/await changes @damianh can you remember if this was left for a reason?

Probably makes sense to be:

````
public override Task<object> Invoke(DynamicDictionary parameters, CancellationToken cancellationToken)
{
    return this.Action.Invoke(parameters, cancellationToken);
 }
````

To avoid the state machine overhead.

**Comment by Luo-Liang** [comment]
Thanks for responding :)

**Comment by damianh** [comment]
Doesn't ring a bell @grumpydev. Probably an oversight.

A PR with this change is welcome @Luo-Liang . As @grumpydev mentions, just return the task, `await`ing not required.

**Comment by thecodejunkie** [comment]
Closed by #2684
